all: add support for conditional fields

a-nogikh opened the pull request:
Closes #2226

**WIP**

Support conditional fields with the following syntax:
```
struct {
  mask   int32
  field  some_substruct    (if[value[mask] & SOME_CONST])
  field2 other_substruct   (if[value[mask] & SOME_CONST == 0])  
}
```

In the example above, `field` will only be included if `mask & SOME_CONST != 0`. In the other case, there'll be `field2`.

The code supports however complex expressions that can be made by combining `==`, `!=`, `&` and `value[path:to:field]`.

It should be quite easy to add more operations later. Also, the way it's implemented now will let us add something like `eq[value[other_field]+1]` for cases when we want one field's value to be dependent on other fields' values. If we ever need it, of course.

The underlying implementation relies on conditional fields for unions. But these can be used independently from conditions on struct fields.
```
s  {
  mask   int32
  f1     u
}

u [
  field  some_substruct    (if[value[parent:mask] & SOME_CONST])
  field2 other_substruct   (if[value[parent:mask] & OTHER_CONST])  
  other third_struct
}
```

The union must either have no fields with conditions, or all fields except the last one must have some condition. During fuzzing, syzkaller would pick the first field that matches its conditions (thanks to the limitation in the previous sentence, there will always be a solution).

TODO:
- [x] Add documentation
- [x] Rewrite some descriptions
- [x] Add more `prog` tests

a-nogikh commented:
Okay, we actually already use `nil` in `prog/encoding.go`.

For normal arguments, it de facto means "generate the default value":
https://github.com/google/syzkaller/blob/28b24332d95f2f7df44ec7e7a5e0025bcadc6277/prog/encoding.go#L411-L413

We even test for it:
https://github.com/google/syzkaller/blob/28b24332d95f2f7df44ec7e7a5e0025bcadc6277/prog/encoding_test.go#L222-L225

Also the `vm` type can have it as a value (that's when `typ == nil` in `encoding.go`). This use does not contradict the purpose of the PR though.
https://github.com/google/syzkaller/blob/28b24332d95f2f7df44ec7e7a5e0025bcadc6277/prog/encoding_test.go#L323-L324

I'm leaning towards introducing some `EmptyArg` in `prog/prog.go` and using the `empty` keyword to signify this in programs. It will fit more nicely into the existing `prog` design and there won't be tons of `if arg == nil` checks. Progs would look like

```
test$conditional_struct(0x4, empty, 0) 
```

dvyukov commented:
Will it be simpler if we lower this to a varlen union with a value and a void field (similar to the current opt template)? Then we can just select the @void field to delete it.

And if we implement unions with conditions on options and expose it in syzland as well, then this will resolve another feature request IIRC (did we want it somewhere?).

a-nogikh commented:
> Will it be simpler if we lower this to a varlen union with a value and a void field (similar to the current opt template)?

What exactly do you mean by "lower"?

**A.** If we keep the syntax for structs as initially proposed here
```
A {
    f1   t1
    f2   t2   (if[val[f1] == 0])
} [packed]
```

Then we'd need to somehow create a fake union type in `pkg/compiler` and substitute it as a field type, as if:
```
autogenerated1 [
  f2 t2 (if[val[parent:f1] == 0]) # we don't support `parent:field` now, but let's imagine
] [varlen]

A {
    f1   t1
    f2   autogenerated1
} [packed]
```

* All field references go via `parent:`, which is not supported yet.
* For `prog`, it will be union, so all errors and restrictions (e.g. about resources) will be somewhat confusing.

**B.** Don't support `if[expr]` attr for struct fields (only for union fields). Use conditional fields via a template explicitly.

Built-in:
```
type optField[T, E] [
	val	T (if[E])
	void	void
] [varlen]
```

In descriptions:
```
A {
    f1   t1
    f2   optField[t2, val[parent:f1] == 0]
} [packed]
```

* We still have to carry on `parent:` (and support it).
* The syntax gets somewhat more clumsy.

**Regarding `if[]` for union fields in general:**

We must select exactly one union field, so when we generate a union with field conditions, we must set the values of depend-on fields in such a way that at least one field can be selected. This is doable, but might be quite tricky (especially when we support more expression operators), so we either do that, or we require that one union field remains without a condition.

In the latter case, though it covers exactly the use case we discuss here, it is unlikely to be very useful for anything else.

And even if we support unions with conditions for all fields.. I don't know where exactly we can apply it. If all field conditions are mutually exclusive, it's no different from a struct with field conditions. Otherwise, if conditions of several union fields match, was the use of such a union with field conditions a good choice in the first place?

---

If we support per-field conditions in unions, we can indeed use it to implement conditions on struct fields, but some parts of the implementation will become more tricky than now (take care about `parent:a:b:c` field references and deal with a more complicated generation/mutation).

If we only support per-field conditions in structs, it will reduce the need for field conditions in unions.

---

Did you have any other options in mind?

dvyukov commented:
I meant A. Syntax and internal representation don't need to match. We already do a number of such transformations in the compiler.

> For prog, it will be union, so all errors and restrictions (e.g. about resources) will be somewhat confusing.

All errors are printed by the compiler before the transformation.
Don't we want the same restrictions about resources? The field may be missing, so we can't assume that the syscall will create the resource.

> We must select exactly one union field, so when we generate a union with field conditions, we must set the values of depend-on fields in such a way that at least one field can be selected. 

I think we should require a "default" field w/o a condition at the end.
Consequently this should be lowered to:

```
autogenerated1 [
  f2 t2 (if[val[parent:f1] == 0]) # we don't support `parent:field` now, but let's imagine
  v void
] [varlen]
```

> All field references go via parent:, which is not supported yet.

IIRC we support `parent:f1` and `parent:parent:f1` and `A:f1`.

a-nogikh commented:
Thanks for the clarification, I'll try to do it this way.

> IIRC we support parent:f1 and parent:parent:f1 and A:f1.

For `parent:f1` -- no, we don't.
https://github.com/google/syzkaller/blob/fb427a0782000106c62de76d251e5a02de5406a9/pkg/compiler/testdata/errors2.txt#L149-L151

a-nogikh commented:
I've rewritten the code to use unions underneath. There are still TODOs (including adding more tests, documentation and rewriting some of the existing descriptions).

@dvyukov  could you please take a look?

---

Now `len` also supports the `parent:field` notation, though it's somewhat tricky syntactically now.

For structs, we used to refer to the whole struct as `parent`, so `parent:field` must logically mean a field in the same structure, just as if `parent:` was not there.
```
X {
 f1  array[int32]
 f2  len[parent:f1, int32]
}
```

For unions, there are no sibling fields, so `parent:other-union-field` is meaningless, so `parent` may refer to the wrapping structure. We could save
```
X {
 f1 array[int32]
 f2 Y
}

Y [
 u1 len[parent:f1, int32]
]
```

`prog/size.go` did not initialize `len`-like fields inside unions by the way, even though we used it in the descriptions at least once:

```
type ebt_among_info_offset[FIELD] [
	offset	offsetof[ebt_among_info:FIELD, int32]
	zero	const[0, int32]
]
```

This should be working well after these changes.

dvyukov commented:
> For structs, we used to refer to the whole struct as parent, so parent:field must logically mean a field in the same structure, just as if parent: was not there.

Yes, this is nasty. In this case, shouldn't it be `parent:parent:field` to make it consistent?

But also when lowering other constructs, we can just use the parent name explicitly:
```
X {
 f1 array[int32]
 f2 Y
}

Y [
 u1 len[X:f1, int32]
]
```

So perhaps we don't need any extensions for field references for this.

dvyukov commented on sys/test/test.txt:
```diff
@@ -486,6 +486,37 @@ offsetof0 {
+}
+
+parent_union [
+	u1	len[parent:f1, int32]
```
The commit description says:
```
Let len[parent:parent:a, T] refer to the structure that encloses the
current one.
```

But this uses just `parent:f1`.

So it's not `parent:parent`, right?
`parent:parent` looks more consistent. Otherwise `parent` in `parent` and in `parent:f` refers to different structs.

a-nogikh replied:
I've adjusted the code to use `parent:parent:field` for unions as well.

dvyukov commented on pkg/compiler/testdata/errors2.txt:
```diff
@@ -123,12 +123,12 @@ s8 {
 
 u0 [
-	f	len[f1, int32]	### len target f1 does not exist
+	f	len[f1, int32]	### len target f1 does not exist in u0, parents: <foo$202>, s7
```
Is the list of parents useful?
I am trying to understand when it would be useful for me, but don't see yet. We print the struct/syscall name where we are looking for the field. I can see it may be useful, esp for "syscall". But I am not sure about parents.

a-nogikh replied:
Hmm, indeed. I've left the `in %s` part, but dropped `parents:`.

dvyukov commented on pkg/compiler/testdata/errors2.txt:
```diff
@@ -144,17 +144,17 @@ foo$201(a ptr[in, s1])
-	f2 len[slen21:a, int32]		### len target a does not exist
-	f3 len[f0:parent, int32]	### parent can't be part of path expressions
+	f1 len[parent:f0, int32]
+	f2 len[slen21:a, int32]		### len target a does not exist in slen2, parents: <foo$204>, slen1
```
How did "slen2" come up here? The parent name is "slen1", and the expression references "slen21". I am confused.

a-nogikh replied:
`slen2` is the actual name of the struct where it was looking for the field named `a`, `slen21` was of type `slen2`

dvyukov commented on pkg/compiler/consts.go:
```diff
@@ -247,16 +271,28 @@ func (comp *compiler) patchConsts(consts0 map[string]uint64) {
 						comp.patchTypeConst(attr.Args[0], consts, &missing)
 					}
 				}
+				foreachFieldAttr(n, func(attr *ast.Type, desc *attrDesc) {
```
Is it possible to use the generic walking to reach all expression leafs instead of these special foreachFieldAttr/ForeachLeafType?
Walking is intended to reach all Type's that are present in the tree. So possible we don't need to do anything special here if we fix walking.

a-nogikh replied:
I've replaced `ForeachLeafType` with `ast.Recursive` (I had to make `ast.Recursive`'s callback return a bool to control whether we go into the ast subtree).

`foreachFieldAttr` is now `foreachFieldAttrConst`, which is just a helper function. I don't think we need generic walking in there.

dvyukov commented on prog/types.go:
```diff
@@ -81,6 +82,62 @@ func (f *Field) Dir(def Dir) Dir {
+	Clone() ValueExpression
+}
+
+type BinaryOperator int64
```
s/int64/int/
Int is fine default type.

dvyukov commented on prog/types.go:
```diff
@@ -81,6 +82,62 @@ func (f *Field) Dir(def Dir) Dir {
 	return def
 }
 
+type ValueExpression interface {
```
I would call it just Expression.
We don't have other expressions and it's assumed expressions evaluate to values.

dvyukov commented on prog/types.go:
```diff
@@ -81,6 +82,62 @@ func (f *Field) Dir(def Dir) Dir {
+	OperatorBinaryAnd
+)
+
+type BinaryOperation struct {
```
s/BinaryOperation/BinaryExpression/
This is an expression.

dvyukov commented:
Nice!
I think in general this is the right approach and it will provide a good base for implementation of other features on top.

We also need some user docs with details and examples.
In particular, I am interested in the alignment details for struct/union fields. This may also need some additional tests for alignment, I see lots of potential for bugs there.

Are conditional fields supported for bitfields? I think we should explicitly prohibit it for now.

Exact rules for union conditional fields.

Expression types story. == and != produce bool result, while & produce int result. Unclear if the end type of the expression is required to be bool.

a-nogikh commented on pkg/compiler/testdata/errors2.txt:
```diff
@@ -446,3 +446,45 @@ foo$non_minimizable_via_union(a ptr[in, union_non_minimizable])	(no_generate)	##
+	f4	int32
+}
+
+some_nested_flags {
```
I've simplified the error descriptions and added successful cases to `all.txt`.

a-nogikh commented on pkg/compiler/attrs.go:
```diff
@@ -23,17 +31,18 @@ type attrDesc struct {
 	attrVarlen     = &attrDesc{Name: "varlen"}
-	attrSize       = &attrDesc{Name: "size", HasArg: true}
-	attrAlign      = &attrDesc{Name: "align", HasArg: true}
+	attrSize       = &attrDesc{Name: "size", Type: intAttr}
```
I think that ultimately we definitely want this, we'll just need some extra code to make sure that no fields are referenced in these const expressions and to evaluate them right in `pkg/compiler`.

I've added a TODO.

a-nogikh commented on pkg/ast/ast.go:
```diff
@@ -252,15 +252,36 @@ func (n *Int) GetName() string {
+type Operator int
+
+const (
+	_ Operator = iota
```
You mean the `_` entry?
I think that by starting from 0 we risk treating a field that was initialized by default as the first enum evalue. Since we are here explicitly doing switches over operators, starting from 1 will help us detect more bugs in the code that deals with these operators.

I see that we can achieve the same effect by `= iota + 1`  though.

dvyukov replied:
Yes, `_` entry.
Let's do `= iota + 1`.

a-nogikh commented:
> In particular, I am interested in the alignment details for struct/union fields. This may also need some additional tests for alignment, I see lots of potential for bugs there.

What would be some examples that come to your mind? Now that we rely on unions, the existing code should already handle complex unions?

> Are conditional fields supported for bitfields? I think we should explicitly prohibit it for now.

Good point. Technically, it should not be too difficult to support them, shouldn't it? We just need to truncate ConstArg's value by `TypeBitSize()` of it's type, right?

a-nogikh commented:
I've tried to rewrite some of the wifi descriptions.
* In some cases it looks convenient if conditional fields could depend on other conditional fields values. If the referenced field is not included, then the dependent fields are also not included. But the implementation might be tricky since paths in `value[]` would be referencing union fields.
* It might make sense to support `&&` and `||` operators right away.

a-nogikh commented:
I've run some side-to-side comparisons with the current upstream version and the version from this PR seemed ~20% slower (apparently due to extra computations we now perform).

I've optimized some conditional fields-related code and added the `prog: cache parents map` commit that also affects existing `len[X, Y]` calculations. Now this version seems to be even faster than the upstream one.

dvyukov commented:
> What would be some examples that come to your mind? 

Simple example:

```
foo {
  a int8
  b int32 [if a == 0]
  c int8
}
```

If b is present, it's aligned to 4 and there is 3 bytes of padding before it.
If b would not be present in the description, then there would be no alignment padding and c is placed right after a.
Presence of b also affects foo's alignment (how it's laid out in a parent struct).

With conditional fields: if the field is missing, does it still have the alignment effect on subsequent fields and the parent, or not? What behavior do we want? What behavior is implemented now? Either way would benefit from a test.

Similar example with unions: if we a non-varlen union with conditional fields, do missing fields contribute to the total union size/alignment?

dvyukov commented:
> > In particular, I am interested in the alignment details for struct/union fields. This may also need some additional tests for alignment, I see lots of potential for bugs there.
> 
> What would be some examples that come to your mind? Now that we rely on unions, the existing code should already handle complex unions?
> 
> > Are conditional fields supported for bitfields? I think we should explicitly prohibit it for now.
> 
> Good point.
> 
> * In terms of making a condition depend on a bitfield value, there should be no problem.
> * Placing a condition on a bitfield is more interesting. Do we support unions that take less than 1 byte size?

I don't think do. All bitfields also must be consecutive fields, with the union they become non-consecutive.
There are also lots and lots of corner cases related to bitfields. I think the most corner cases across all features we have. Each line here is a separate corner case:
https://github.com/google/syzkaller/blob/master/sys/test/test/bf
https://github.com/google/syzkaller/blob/master/sys/test/test/bf2
https://github.com/google/syzkaller/blob/master/sys/test/test/bf2_be
https://github.com/google/syzkaller/blob/master/sys/test/test/bf_be
https://github.com/google/syzkaller/blob/master/sys/test/exec.txt#L98-L248
So I would just prohibit conditional bitfields on the compiler level for now, if we don't need them right now.

dvyukov commented:
> I've run some side-to-side comparisons with the current upstream version and the version from this PR seemed ~20% slower (apparently due to extra computations we now perform).
> 
> I've optimized some conditional fields-related code and added the `prog: cache parents map` commit that also affects existing `len[X, Y]` calculations. Now this version seems to be even faster than the upstream one.

Btw, for this type of this things you can use Go's microbenchmarks, e.g.:


```
$ go test -run=none -bench=BenchmarkMutate -benchtime=0.5s -count=10 ./prog | tee /tmp/1
BenchmarkMutate-72    	    5335	     98306 ns/op	  378774 B/op	    3270 allocs/op
BenchmarkMutate-72    	    6972	     93111 ns/op	  423663 B/op	    3320 allocs/op
BenchmarkMutate-72    	    6625	     89595 ns/op	  423379 B/op	    3319 allocs/op
...

$ go test -run=none -bench=BenchmarkMutate -benchtime=0.5s -count=10 ./prog | tee /tmp/2
...

$ benchstat /tmp/1 /tmp/2
          │   /tmp/1    │            /tmp/2             │
          │   sec/op    │   sec/op     vs base          │
Mutate-72   93.67µ ± 1%   97.80µ ± 5%  ~ (p=0.052 n=10)

          │    /tmp/1    │             /tmp/2             │
          │     B/op     │     B/op      vs base          │
Mutate-72   413.6Ki ± 1%   413.2Ki ± 1%  ~ (p=0.165 n=10)

          │   /tmp/1    │               /tmp/2               │
          │  allocs/op  │  allocs/op   vs base               │
Mutate-72   3.325k ± 0%   3.320k ± 0%  -0.15% (p=0.026 n=10)
```

a-nogikh commented:
Benchstat:

* `/tmp/upstream` is the current syzkaller HEAD.
* `/tmp/cond-fields` are all base changes from this PR.
* `/tmp/cache-map` is `/tmp/cond-fields` + [prog: cache parents map](https://github.com/a-nogikh/syzkaller/commit/2e5ff016e7660707b452942d01eafc646a508570)
* `/tmp/parent-stack` is `prog: collect parents during arg traversal`

I don't know how suitable that particular benchmark is though. The vast majority of syzkaller executions are minimiziations (and that's presumably where we recalculate conditional fields and lengths most of the time).

```
          │ /tmp/upstream │          /tmp/cond-fields          │            /tmp/cache-map             │          /tmp/parent-stack          │
          │    sec/op     │   sec/op     vs base               │   sec/op     vs base                  │   sec/op     vs base                │
Mutate-36     139.9µ ± 1%   149.6µ ± 5%  +6.89% (p=0.000 n=15)   152.7µ ± 4%  +9.14% (p=0.000 n=15+10)   163.3µ ± 4%  +16.71% (p=0.000 n=15)

          │ /tmp/upstream │           /tmp/cond-fields           │             /tmp/cache-map              │          /tmp/parent-stack           │
          │     B/op      │     B/op      vs base                │     B/op      vs base                   │     B/op      vs base                │
Mutate-36    403.8Ki ± 1%   505.9Ki ± 2%  +25.27% (p=0.000 n=15)   509.0Ki ± 2%  +26.04% (p=0.000 n=15+10)   503.2Ki ± 1%  +24.60% (p=0.000 n=15)

          │ /tmp/upstream │          /tmp/cond-fields          │            /tmp/cache-map             │          /tmp/parent-stack          │
          │   allocs/op   │  allocs/op   vs base               │  allocs/op   vs base                  │  allocs/op   vs base                │
Mutate-36     3.286k ± 0%   3.427k ± 1%  +4.29% (p=0.000 n=15)   3.433k ± 1%  +4.47% (p=0.000 n=15+10)   4.506k ± 1%  +37.13% (p=0.000 n=15)
```

a-nogikh commented:
I've prohibited conditions on bitfields.

Regarding unions:

```
foo {
  a int8
  b int32 (if[value[a] == 0])
  c int8
}
```

I've adjusted the code to treat fields with conditions as varlens, so if such a field appears in the middle of a struct, the struct must be `[packed]`. If a struct has a conditional field (anywhere), it's assumed to be varlen.

> With conditional fields: if the field is missing, does it still have the alignment effect on subsequent fields and the parent, or not? What behavior do we want? What behavior is implemented now? Either way would benefit from a test.

So now we're only considering the case of `[packed, align[N]]` structs -- that is, syzkaller should pad the total size of the structure to a multiple of `N`. I can add a few such tests (one with a single struct, one with a nested one) to `sys/test/test`.

> Similar example with unions: if we a non-varlen union with conditional fields, do missing fields contribute to the total union size/alignment?

As far as I see, there should be no difference for unions. Whether there are conditions or there are no conditions, we anyway assume it may take one of the options.

dvyukov commented:
> Maybe we really don't need that optimization. Though it's really weird to see that these seemingly reasonable changes made the benchmark even slower.

Maybe we need a different, better benchmark. I used Mutate only as an existing example.

This may be a reason for performance degradation:
```
allocs/op   vs base
4.506k ± 1%  +37.13%
```
memory allocations are bad for performance in garbage-collected languages. They trigger garbage collections that are relatively slow.

a-nogikh commented:
After https://github.com/google/syzkaller/commit/8a70020f31559cb43209df6788c90da90f05cec2 and some performance optimizations of `prog: collect parents during arg traversal` (only collect parents in size.go and expr.go):

```
          │ /tmp/upstream │  /tmp/optimized-parent-slice  │
          │    sec/op     │   sec/op     vs base          │
Mutate-36     117.0µ ± 6%   115.2µ ± 4%  ~ (p=0.653 n=15)

          │ /tmp/upstream │     /tmp/optimized-parent-slice     │
          │     B/op      │     B/op      vs base               │
Mutate-36    403.4Ki ± 1%   375.0Ki ± 2%  -7.05% (p=0.000 n=15)

          │ /tmp/upstream │    /tmp/optimized-parent-slice     │
          │   allocs/op   │  allocs/op   vs base               │
Mutate-36     3.273k ± 0%   3.435k ± 0%  +4.95% (p=0.000 n=15)
```

a-nogikh commented:
* Added some alignment tests (they worked well right away).
* Added minimization tests (they discovered a bug that I then fixed).
* Added documentation.

dvyukov commented on docs/syscall_descriptions_syntax.md:
````diff
@@ -405,6 +411,155 @@ foo(a ptr[in, array[int8, MY_PATH_MAX]])
 define MY_PATH_MAX	PATH_MAX + 2
 ```
 
+## Conditional fields
````
@tarasmadan please review the user documentation with fresh view.
I already know what this is doing, so can't review as a new user would read it.

dvyukov commented on prog/expr.go:
```diff
@@ -0,0 +1,195 @@
+	r.inPatchConditional = true
+	defer func() { r.inPatchConditional = false }()
+
+	var extraCalls []*Call
```
If you gave names to return values, then you can just use them in the function and have an empty `return` in the end.
This avoids additional variable declarations and make it impossible to assign to a wrong variable (currently it's still possible to do `changed = true` which will be lost).

dvyukov commented on prog/expr.go:
```diff
@@ -0,0 +1,195 @@
+	return nil
+}
+
+var ErrViolatedConditions = errors.New("conditional fields rules violation")
```
unexport this

dvyukov commented on prog/types.go:
```diff
@@ -81,6 +82,68 @@ func (f *Field) Dir(def Dir) Dir {
 	return def
 }
 
+type ArgFinder func(path []string) Arg
```
Do ArgFinder, SquashedArgFound and Expression need to be exported from the package?

dvyukov commented on prog/rand.go:
```diff
@@ -582,8 +583,9 @@ func (r *randGen) generateParticularCall(s *state, meta *Syscall) (calls []*Call
 	}
 	c := MakeCall(meta, nil)
 	c.Args, calls = r.generateArgs(s, meta.Args, DirIn)
+	moreCalls, _ := r.patchConditionalFields(c, s)
```
What are the rules for conditional fields vs len type?
If we call patchConditionalFields before assignSizesCall, I assume sizes are not assigned when we evaluate conditional fields, does it mean that len fields cannot be used in conditional field expressions?
Similarly, does len of a struct that contains conditional fields work?

a-nogikh replied:
Conditional fields may only reference `typeInt` fields in their expressions, so there's no need to set sizes before we call `patchConditionalFields()`. But sizes do depend on them, so they should only be calculated after we have ensured that conditional fields are set correctly.

> Similarly, does len of a struct that contains conditional fields work?

If `len` works fine with structs that contain `optional[T]`, then it should also work with conditional fields -- the underlying implementation is similar. I can add some tests just in case.

dvyukov replied:
> Conditional fields may only reference typeInt fields in their expressions

I see.

> If len works fine with structs that contain optional[T], then it should also work with conditional fields

But if sizes are calculated before patchConditionalFields, does it mean that size will always include the field even if it will be later removed?

a-nogikh replied:
Before `patchConditionalFields()` is called, some conditional fields may be inconsistent (e.g. because we mutated some int field) or may not yet have been calculated at all (e.g. because we've just generated the type). If we calculate sizes on such a state, the assigned sizes will likely be just incorrect.

dvyukov replied:
So what is the size of such structs? Is it random?
Let's capture this in the docs.

a-nogikh replied:
I wouldn't say it's random -- it'd be a bug in syzkaller code if lens are not updated after conditional fields are patched.

I've added a few `size_test.go` tests.

dvyukov commented:
Add this to pkg/compiler.TestFuzz ;)
```
type r[N][
R	r[t[r[t[N,NN,NN,N,T]],N,N,N,T]]
]
n(a r[t])
```

xairy commented on docs/syscall_descriptions_syntax.md:
````diff
@@ -405,6 +411,155 @@ foo(a ptr[in, array[int8, MY_PATH_MAX]])
+some_call(&AUTO={{AUTO, 0x0}, @void, []})
+```
+
+A structure that has conditional fields is assumed to be `[varlen]` and.  If
````
Typo with `and.` here.

xairy commented on docs/syscall_descriptions_syntax.md:
```diff
@@ -405,6 +411,155 @@ foo(a ptr[in, array[int8, MY_PATH_MAX]])
+
+alternatives [
+  int     int64 (if[value[struct:type] == 0x1])
+  arr     array[int64, 5] (if[value[struct:type] == 0x1])
```
Perhaps, it makes sense to use `== 0x2` here to better demonstrate the idea?

xairy commented on docs/syscall_descriptions_syntax.md:
````diff
@@ -405,6 +411,155 @@ foo(a ptr[in, array[int8, MY_PATH_MAX]])
+```
+
+Syzkaller will always obey these restrictions. Specifically, it will select the
+first union field whose condition is satisfied.
````
Hm, wouldn't it make more sense to randomly select one field out of the ones that satisfy the condition if there are many of them?

a-nogikh replied:
This semantics was necessary to implement conditional struct fields on top of conditional union fields.

If we go with `take random union field that matches`, then the field without condition may always be taken, so the current transformation:

`field type (if[A])` =>
```
[
value type (if[A])
void void
]
```

no longer fully works. If we transform to (*)
```
[
value type (if[A])
void void (if[(A) == 0])
]
```

then we need to support unions with conditions on all fields. Then:
* We have trouble designating the default union field. 
* (Specifically) We must be able to prove that at least one of the conditions is always satisfiable. It can be tricky.

So the current semantics was just simpler.

Theoretically we could do the following though:
* For user-defined unions, continue to require the last option to have no if condition.
* Transform structs with conditional fields as in (*), i.e. so that all union options do have conditions. We know that one of the conditions is always true, so it will work.

xairy replied:
I don't get what you meant in the part about transformations, but if a union field is now selected randomly - that's perfect!

xairy commented on docs/syscall_descriptions_syntax.md:
````diff
@@ -405,6 +411,155 @@ foo(a ptr[in, array[int8, MY_PATH_MAX]])
+```
+sub_struct {
+  f0 int
+  f1 int (if[value[struct:f2]]) # Reference a field in a parent struct.
````
This condition is the same as `value[struct:f2] != 0`, right? Makes sense to clarify.

a-nogikh replied:
Added a note. Thanks!

xairy replied:
I don't see the note...

a-nogikh replied:
Hmm, strange, I remember typing it..

Fixed.

a-nogikh commented:
> `type r[N][
> R	r[t[r[t[N,NN,NN,N,T]],N,N,N,T]]
> ]
> n(a r[t])
> `

It also happens on the upstream revision. I've filed a separate issue: https://github.com/google/syzkaller/issues/4434

dvyukov reviewed (commented):
Looks good to me, will leave to @tarasmadan to review docs with fresh eye.

tarasmadan commented on docs/syscall_descriptions_syntax.md:
```diff
@@ -405,6 +411,155 @@ foo(a ptr[in, array[int8, MY_PATH_MAX]])
+
+packet {
+  header  header_fields
+  integer int64  (if[value[header:haveField] == 0x1])
```
haveField or haveInteger?

tarasmadan commented on docs/syscall_descriptions_syntax.md:
````diff
@@ -405,6 +411,155 @@ foo(a ptr[in, array[int8, MY_PATH_MAX]])
+In this example, the `packet` structure will include the field `integer` only
+if `header.haveInteger == 1`. That corresponds to e.g. the following program:
+```
+some_call(&AUTO={{AUTO, 0x1}, @value=0xabcd, []})
````
Reading only this page from the beginning I don't understand this line.
Could you please simplify the example of add some additional references?

a-nogikh replied:
I've added tables with `packet` layouts.

tarasmadan commented on docs/syscall_descriptions_syntax.md:
````diff
@@ -405,6 +411,155 @@ foo(a ptr[in, array[int8, MY_PATH_MAX]])
+```
+
+If `header.haveInteger` is not `1`, syzkaller will just pretend that the field
+`integer` does not exist -- much like it's done for `optional[T]`.
````
It is not clear what is optional[T].

tarasmadan reviewed (commented):
Every comment reflects my understanding of the material from the beginning of this doc to the comment line. It makes sense to read them all first.

xairy commented on docs/syscall_descriptions_syntax.md:
````diff
@@ -405,6 +411,155 @@ foo(a ptr[in, array[int8, MY_PATH_MAX]])
+some_call(&AUTO={0x0, @default=0xabcd})
+```
+
+The following definition would fail to compile:
````
I would say `Thus, the following ...` to connect it to the previous statement.

xairy commented on docs/syscall_descriptions_syntax.md:
````diff
@@ -405,6 +411,155 @@ foo(a ptr[in, array[int8, MY_PATH_MAX]])
+] [varlen]
+```
+
+Syzkaller will always obey these restrictions. During prog mutation and generation
````
I would also drop `Syzkaller will always obey these restrictions`: I'm not sure what additional meaning it adds.

a-nogikh commented on docs/syscall_descriptions_syntax.md:
````diff
@@ -405,6 +411,155 @@ foo(a ptr[in, array[int8, MY_PATH_MAX]])
+```
+sub_struct {
+  f0 int
+  f1 int (if[value[struct:f2]]) # Reference a field in a parent struct.
````
> was used to name the field parent struct ("parent" param afair)

Could you please clarify where exactly?

a-nogikh commented:
Judging by a `syz-testbed` experiment, the [prog: collect parents during arg traversal](https://github.com/google/syzkaller/pull/4380/commits/889168b8b564a982b0bcb7ef3306708c66d7c0fb) commit has introduced some minor fuzzing efficiency regression: syzkaller finds a bit less signal and coverage than without this commit.

a-nogikh commented on docs/syscall_descriptions_syntax.md:
````diff
@@ -181,10 +184,13 @@ Unions are described as:
 "]" ("[" attribute* "]")?
 ```
 
+You may specify conditions that determine what field must be selected. If they are not
````
Made it clearer.

a-nogikh commented on docs/syscall_descriptions_syntax.md:
```diff
@@ -405,6 +411,155 @@ foo(a ptr[in, array[int8, MY_PATH_MAX]])
+  f0 int
+  f1 int (if[value[f0] == 0x1])
+  f2 int (if[value[f1] == 0x1])
+}
```
Indeed, this example wouldn't compile. Added a `[packed]`.

a-nogikh commented on docs/syscall_descriptions_syntax.md:
```diff
@@ -405,6 +411,155 @@ foo(a ptr[in, array[int8, MY_PATH_MAX]])
+  int     int64 (if[value[struct:type] == 0x1])
+  arr     array[int64, 5] (if[value[struct:type] == 0x1])
+  default int32
+] [varlen]
```
You cannot mark structs `[varlen]` -- I've removed the notion from the document to make it less confusing. Structs must be marked with `[packed]` if they have a variable length field that is not the last field of the structure.

Variable length unions must be marked with `[varlen]`. I think it's done so that the fact is explicitly seen.

a-nogikh commented:
> Judging by a `syz-testbed` experiment, the [prog: collect parents during arg traversal](https://github.com/google/syzkaller/pull/4380/commits/889168b8b564a982b0bcb7ef3306708c66d7c0fb) commit has introduced some minor fuzzing efficiency regression: syzkaller finds a bit less signal and coverage than without this commit.

I've tried to find the cause of the regression by both examining the code and by running `tools/syz-mutate` with different seeds and comparing the results (both with and without `-corpus`), but still no success. So I decided to give up for now and dropped the commit, I'll send it separately after this series is merged.

On a `syz-testbed`, the version that has the `prog: collect parents during arg traversal` commit:
* Reaches 0.6% less coverage and signal.
* Does 2% more `exec fuzz` and `exec gen` calls (so it finds less new coverage in normal operation and the fuzzing queue gets empty more often)
* Does 1.3% more hints while having -1.8% `exec seed`. Does it mean that there are e.g. more zero-valued fields that cause more hint substitution candidates??
* Has -5% `new inputs` and -0.8% `corpus` size.

It's very weird because if it would generate/mutate programs differently, I would have noticed it in the `tools/syz-mutate` output.

xairy commented on docs/syscall_descriptions_syntax.md:
````diff
@@ -405,6 +423,167 @@ foo(a ptr[in, array[int8, MY_PATH_MAX]])
+```
+
+In this case, the specific union field will be selected depending on the value
+of the `type` field. For example, if `type` is `0x1`, then it's `int`:
````
So `default` will never be selected if `type` is `1`? I would say it makes sense to select it sometimes too but with a lower probability. Or does this already happen and this documentation line is outdated?

a-nogikh replied:
>  Or does this already happen and this documentation line is outdated?

It already happens, thank for catching this! Fixed.

a-nogikh commented:
Okay, let's go.
